File: src/main/java/com/amazonaws/iot/fleetwise/vehiclesimulator/cli/PropertiesProvider.java
package com.amazonaws.iot.fleetwise.vehiclesimulator.cli;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.time.Instant;
import java.util.Properties;

import org.jetbrains.annotations.Nullable;

public final class PropertiesProvider {

    private final Properties properties = new Properties();

    public static final PropertiesProvider INSTANCE = new PropertiesProvider();

    private PropertiesProvider() {
        try (InputStream inputStream = PropertiesProvider.class.getResourceAsStream("/version.properties")) {
            String line;
            if (inputStream != null) {
                BufferedReader reader  =new BufferedReader(new InputStreamReader(inputStream));
                while ((line = reader.readLine()) != null) {
                    String[] keyValuePair = line.split(":", 2);
                    properties.put(keyValuePair[0], keyValuePair[1]);
                }
            }
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }

    public String getVersion() {
        return properties.getProperty("version");
    }

    @Nullable
    public Instant getBuildTime() {
        return Instant.ofEpochMilli(Long.parseLong(properties.getProperty("buildTime")));
    }

}

File: src/main/kotlin/com/amazonaws/iot/vehiclesimulator/cli/PropertiesProvider.kt
package com.amazonaws.iot.fleetwise.vehiclesimulator.cli

import picocli.CommandLine
import java.time.Instant
import java.util.Properties

class VersionProvider : CommandLine.IVersionProvider {
    override fun getVersion(): Array<String> {
        return arrayOf(
            "${PropertiesProvider.version} built at ${PropertiesProvider.buildTime}"
        )
    }
}

object PropertiesProvider {
    private val properties: Properties? by lazy {
        javaClass.getResourceAsStream("/version.properties")?.use {
            Properties().apply {
                load(it)
            }
        }
    }

    val version: String by lazy {
        properties?.getProperty("version") ?: "Undefined"
    }

    val buildTime: Instant? by lazy {
        properties?.getProperty("buildTime")?.toLongOrNull()?.let {
            Instant.ofEpochMilli(it)
        }
    }
}
